pkg/storage/utils/decomposedfs/node/locks.go

```diff
@@ -307,3 +312,24 @@ func (n *Node) hasLocks(ctx context.Context) bool {
 	_, err := os.Stat(n.LockFilePath()) // FIXME better error checking
 	return err == nil
 }
+
+func isLockModificationAllowed(oldLock *provider.Lock, newLock *provider.Lock) (bool, error) {
+	if oldLock.Type == provider.LockType_LOCK_TYPE_SHARED {
+		return true, nil
+	}
+
+	var allowed bool = true
+	var err error
+
+	if oldLock.AppName != "" && oldLock.AppName != newLock.AppName {
+		allowed = false
+		err = errtypes.PermissionDenied("cannot change holder (app name) when refreshing a lock")
+	}
+
+	if oldLock.User != nil && !utils.UserEqual(oldLock.User, newLock.GetUser()) {
+		allowed = false
+		err = errtypes.PermissionDenied("cannot change holder (user) when refreshing a lock")
+	}
+
```

Comment:
Not sure if the error logic is correct here: If the oldLock.AppName is empty, the lock modification is allowed, same for user. Is that intended?

Reply:
If the user is not present in both locks, we allow modification. Same applies for the app name.